Code/GraphMol/MonomerMol/AtomisticToMM.cpp

```diff
@@ -0,0 +1,216 @@
+#include "MonomerMol.h"
+
+namespace RDKit
+{
+
+namespace
+{
+
+const std::string ATTACH_FROM{"attachFrom"};
+const std::string MONOMER_IDX1{"monomerIndex1"};
+const std::string MONOMER_IDX2{"monomerIndex2"};
+const std::string REFERENCE_IDX{"referenceIndex"};
+
+// 3-letter to 1-letter amino acid code mapping
+// These are not currently in the monomer database, but some have
+// symbols that are already in the monomer database. We may need to
+// figure out how to have multiple 3-letter codes for a single symbol
+// polymer_type pair (Histidine is the best example)
+const std::unordered_map<std::string, char> backup_res_table({
+    {"ARN", 'R'}, // Neutral-Arginine
+    {"ASH", 'D'}, // Protonated Aspartic
+    {"GLH", 'E'}, // Protonated Glutamic
+    {"HID", 'H'}, // Histidine (protonated at delta N)
+    {"HIE", 'H'}, // Histidine (protonated at epsilon N)
+    {"HIP", 'H'}, // Histidine (protonated at both N)
+    {"HSD", 'H'}, // Histidine (protonated at delta N, CHARMM name)
+    {"HSE", 'H'}, // Histidine (protonated at epsilon N, CHARMM name)
+    {"HSP", 'H'}, // Histidine (protonated at both N, CHARMM name)
+    {"LYN", 'K'}, // Protonated Lysine
+    {"SRO", 'S'}, // Ionized Serine
+    {"THO", 'T'}, // Ionized Threonine
+    {"TYO", 'Y'}, // Ionized Tyrosine
+    {"XXX", 'X'}, // Unknown
+});
+
+struct Linkage {
+    unsigned int monomer_idx1;
+    unsigned int monomer_idx2;
+    unsigned int attach_from;
+    unsigned int attach_to;
+    std::string to_string() const
+    {
+        return "R" + std::to_string(attach_from) + "-R" + std::to_string(attach_to);
+    }
+};
+
+void remove_waters(RDKit::RWMol& mol)
+{
+    mol.beginBatchEdit();
+    auto is_water = [](const RDKit::Atom* atom) {
+        const auto res_info = dynamic_cast<const RDKit::AtomPDBResidueInfo*>(
+            atom->getMonomerInfo());
+        if (res_info && res_info->getResidueName() == "HOH ") {
+            return true;
+        }
+        return false;
+    };
+    for (auto atom : mol.atoms()) {
+        if (is_water(atom)) {
+            mol.removeAtom(atom);
+        }
+    }
+    mol.commitBatchEdit();
+}
+
+void find_chains_and_residues(
+    const RDKit::ROMol& mol,
+    std::map<std::string, std::map<int, std::vector<unsigned int>>>&
+        chains_and_residues)
+{
+    // Find all chains and residues in the molecule
+    for (unsigned int i = 0; i < mol.getNumAtoms(); ++i) {
+        const auto atom = mol.getAtomWithIdx(i);
+        const auto res_info = static_cast<const RDKit::AtomPDBResidueInfo*>(
+            atom->getMonomerInfo());
+        const auto chain_id = res_info->getChainId();
+        const auto res_num = res_info->getResidueNumber();
+        chains_and_residues[chain_id][res_num].push_back(i);
+    }
+}
+
+boost::shared_ptr<RDKit::RWMol>
+annotated_atomistic_to_monomeristic(const RDKit::ROMol& input_mol)
+{
+    // Make RWMol and remove waters
+    RDKit::RWMol mol(input_mol);
+    remove_waters(mol);
```

Comment:
why do we remove the waters?

Reply:
Otherwise they will be turned into monomers on the `MonomerMol`; I guess we could ignore them in the conversion but we are working on a copy anyway so this seems simpler

Technically, I think we want to remove all solvents

Reply:
make sense. I'd hope that behavior would be opt-in since this would mean that the "compression" wouldn't be lossless by default.